test/broadway_test.exs

```diff
@@ -441,6 +457,66 @@ defmodule BroadwayTest do
     end
   end
 
+  describe "handle producer crash" do
+    setup do
+      test_pid = self()
+
+      handle_message = fn message, _ ->
+        send(test_pid, {:message_handled, message})
+        message
+      end
+
+      context = %{
+        handle_message: handle_message,
+        handle_batch: fn _, batch, _, _ -> batch end
+      }
+
+      broadway_name = new_unique_name()
+
+      {:ok, _pid} =
+        Broadway.start_link(ForwarderWithCustomHandlers,
+          name: broadway_name,
+          resubscribe_interval: 0,
+          context: context,
+          producers: [
+            default: [
+              module: ManualProducer,
+              arg: %{test_pid: self()},
+              transformer: {Transformer, :transform, []}
+            ]
+          ],
+          processors: [stages: 1, min_demand: 1, max_demand: 2],
+          publishers: [default: [batch_size: 2]]
+        )
+
+      :ok
+    end
+
+    @tag :capture_log
+    test "producer will be restarted" do
+      assert_receive {:producer_initialized, producer}
+      refute_receive {:producer_initialized, _}
+
+      async_push_messages(producer, [:kill_producer])
+
+      assert_receive {:producer_initialized, ^producer}
```

Comment:
Let's augment this test to also check that the processors, batchers, and consumers are not terminated. There is a similar test for processors that we can mirror.